Specification of timecourses & long condition table (#581)
PEtab-dev/PEtab: doc/v2/documentation_data_format.rst

machine- and human-readable and -writable format that makes use of existing
standards.

.. figure:: gfx/petab_scope_and_files.png
   :alt: A typical setup for data-based modeling studies and its
         representation in PEtab.
   :scale: 80%

   **Figure 1: A typical setup for data-based modeling studies and its
   representation in PEtab.**

Overview
---------

The PEtab data format defines a parameter estimation problem using multiple
text-based files in `Tab-Separated Values (TSV)
<https://www.iana.org/assignments/media-types/text/tab-separated-values>`_
format (Figure 2), including:

- A :ref:`model <v2_model>` file specifying the base model
  [SBML, CELLML, BNGL, PYSB, ...].

- A :ref:`measurement file <v2_measurments_table>` containing experimental
  data used for model fitting [TSV].

- (optional) A :ref:`conditions file <v2_conditions_table>` specifying model
  inputs and condition-specific parameters [TSV].

- (optional) An :ref:`experiments file <v2_experiments_table>` describing
  sequences of experimental conditions applied to the model [TSV].

- An :ref:`observables file <v2_observables_table>` defining the observation
  model [TSV].

> - A :ref:`parameters file <v2_parameters_table>` defining estimable

Reviewer: To note close the door on Bayesians :), I would write 

I think parameters to be estimated or inferred.

> - A :ref:`parameters file <v2_parameters_table>` defining estimable
>   parameters and related information [TSV].

Reviewer 2: ```suggestion
- A :ref:`parameters file <v2_parameters_table>` to set parameter values globally and to specify the parameters to be estimated as well as their parameter bounds and prior distributions [TSV].
```

for example, the targets of SBML assignment rules.

* **Constant entities**: Entities are that are defined in terms of a constant
value but may be subject to event assignments, e.g., parameters of an SBML
model that are not targets of rate rules or assignment rules.

.. _v2_conditions_table:

Conditions table
----------------

The conditions table defines discrete changes to the model. These (sets of)
changes typically represent interventions, perturbations, or changes in the
environment of the system of interest. These modifications are referred to as
(experimental) *conditions*.

Conditions are applied at specific time points, which are defined in the
:ref:`v2_experiments_table`. This allows for the specification of time
courses or experiments spanning multiple time periods. A time period is the
interval between two consecutive time points in the experiments table for a
given experiment or the time between the last time point of an experiment and
the time point of the last measurement for that experiment in the
:ref:`v2_measurements_table`.

The conditions tably only allows changes in the model state, not the model

> structure. That mean that only constant or differential entities (e.g.,

Reviewer 2: ```suggestion
structure. That means that only constant or differential entities (e.g.,
```

| patient_3           |                 0 | condition_2     |
+---------------------+-------------------+-----------------+
| intervention_effect |               -20 | no_lockdown     |
+---------------------+-------------------+-----------------+
| intervention_effect |                20 | mild_lockdown   |
+---------------------+-------------------+-----------------+
| intervention_effect |                40 | severe_lockdown |
+---------------------+-------------------+-----------------+

The order of the rows is arbitrary, but specifying the rows in ascending order
of time may improve human readability.

Detailed field description
~~~~~~~~~~~~~~~~~~~~~~~~~~

The time courses table has three mandatory columns ``experimentId``,
``time``, and ``conditionId``:

- ``experimentId`` [PETAB_ID, REQUIRED]

  A unique identifier for the experiment, referenced by the ``experimentId``
  column in the :ref:`v2_measurements_table`.

- ``time``: [NUMERIC or ``-inf``, REQUIRED]

  The time at which the specified condition will become active, in the time
  unit specified in the model. ``-inf`` indicates that the respective
  condition will be simulated until a steady state is reached.

>   If the simulation of a condition with steady fails to reach a steady state,

Reviewer 2: ```suggestion
  If the simulation of a condition with steady state fails to reach a steady state,
```

  and the condition is required for the evaluation of simulation at
  measurement points, the evaluation of the model is not well-defined.
  In such cases, PEtab interpreters should notify the user, for example, by
  returning ``NaN`` or ``inf`` values for the objective function.
  PEtab does not specify a numerical criterion for steady states.
  Any event triggers defined in the model must also be checked during this
  pre-simulation.

>   Multiple conditions cannot be applied simultaneously within in a single

Reviewer 2: ```suggestion
  Multiple conditions cannot be applied simultaneously within a single
```

  experiment; each ``time`` values must be unique for each ``experimentId``.

>   ``time`` will override any initial time specified in the model.

Reviewer 2: ```suggestion
  ``time`` will override any initial time specified in the model, except in the case of ``time`` =  `ìnf``, in which case the model-specified time will be used (or 0, if the model does not explicitly specify an initial time).
```

- Fields enclosed in "[]" are optional and may be left empty.
- The following data types are used in descriptions below:

  - ``STRING``: Any string.
  - ``NUMERIC``: Any number excluding ``NaN`` / ``inf`` / ``-inf``
  - ``MATH_EXPRESSION``: A mathematical expression according to the
    `PEtab math expression syntax <math_expressions>`_.
  - ``PETAB_ID``: A string that is a valid PEtab ID.
  - ``NON_PARAMETER_TABLE_ID``: A valid PEtab ID referring to a constant or
    differential entity (:ref:`v2_model_entities`), including PEtab output
    parameters, but excluding parameters listed in the
    :ref:`v2_parameters_table` (independent of their ``estimate`` value).

Changes from PEtab 1.0.0
------------------------
PEtab 2.0.0 is a major update of the PEtab format. The main changes are:

* Support for models in other formats than SBML (:ref:`v2_model`).
* Revised condition table format from wide to long format
  (:ref:`v2_conditions_table`).
* ``simulationConditionId`` and ``preequilibrationConditionId`` in the
  :ref:`v2_measurements_table` are replaced by ``experimentId`` and a more
  flexible way for defining experiments and time courses

>   (:ref:`v2_experiments_table`).

Reviewer 3: And also now optional, if the user wants to simulate the SBML model as-is

  aspects, including overriding values via the condition table
  (:ref:`v2_math_expressions`).
* Support for format extensions.
* Observable IDs can now be used in observable and noise formulas
  (:ref:`v2_observables_table`).

.. _v2_model:
.. _v2_model_entities:

Model definition
----------------

PEtab 2.0.0 is **model format agnostic**, meaning it does not depend on a
specific model description. The model file is referenced in the :ref:`PEtab
problem description (YAML) <v2_problem_yaml>` by its file name or a URL.

PEtab distinguishes between three types of entities:

* **Differential entities**: Entities that are defined in terms of a
  time-derivative, e.g., the targets of SBML rate rules or species that change
  due to participation in reactions (reactants or products).

* **Algebraic entities**: Entities that are defined in terms of an algebraic
  assignment rather than time derivatives. They are not necessarily constant,
  for example, the targets of SBML assignment rules.

* **Constant entities**: Entities are that are defined in terms of a constant
  value but may be subject to event assignments, e.g., parameters of an SBML

>   model that are not targets of rate rules or assignment rules.

Reviewer 3: It's unclear that algebraic assignments like SBML initialAssignmentRules are constant entities. I'm not sure how to best word that here. In PEtab SciML there is an attempt at distinguishing these by calling this "prior to simulation". I guess the difference between algebraic entities and constant entities is that constant entities are piecewise constant (e.g. initialAssignmentRules or eventAssignments can affect them, but otherwise they are constant), and algebraic entities are everything else.

Reviewer 3: The thing that is unclear to me is that both of these are algebraic, but one is compatible with "constant entity" whereas the other is not:
- initialAssignmentRule is an algebraic equation that is evaluated at one time point
- assignmentRule is an algebraic equation that is evaluated throughout time

Reviewer 2: I tried to rephrase it, please check again.

Reviewer 3: Looks good, the intention of the different entities is clear now :+1: Thanks

.. _v2_conditions_table:

Conditions table
----------------

> The conditions table defines discrete changes to the model. These (sets of)

Reviewer 3: To match the start of the Experiments table section

```suggestion
The optional conditions table defines discrete changes to the model. These (sets of)
```

changes typically represent interventions, perturbations, or changes in the
environment of the system of interest. These modifications are referred to as
(experimental) *conditions*.

Conditions are applied at specific time points, which are defined in the
:ref:`v2_experiments_table`. This allows for the specification of time
courses or experiments spanning multiple time periods. A time period is the
interval between two consecutive time points in the experiments table for a
given experiment or the time between the last time point of an experiment and
the time point of the last measurement for that experiment in the

> :ref:`v2_measurements_table`.

Reviewer 3: Since this is a user's choice, I don't think we need to specify that the simulation stops at the last measurement. A tool might want to give users an easier way to perform predictions, than providing a dummy measurements table.

Reviewer 2: Change to "between the last time point and the end of the simulation" ?

Reviewer 2: Updated.

Reviewer 4: I could imagine adding an empty time interval at the end without any condition with the time>last timepoint. I.e. one could have a time interval at the end where no measurements exist. In this case the simulation has to run until the last specified time point of the experiment.
This should be clarified and would provide a simple manner to encode the end of the simulation if one wants, e.g. for visualization or other purposes.

Reviewer 2: I'd consider that out of scope here. For the purpose of parameter estimation, tools can just stop at the last measurement timepoint.

Your use case would have to be addressed in some visualization function. For proper visualization, additional information such as the output timepoints would be required. I don't think we should try to get that in here.

Reviewer 4: Or could I just add a dummy timepoint to the measurement table with a NaN value at the end?

Reviewer 2: > Or could I just add a dummy timepoint to the measurement table with a NaN value at the end?

That would currently be considered an error. NaN is not allowed. We had that implemented at some point, but it's an abuse of the measurements table. We can continue that discussion outside this PR if you'd consider it important to change that.

If the only purpose is visualization, I'd rather specify the endpoint and potentially time-resolution in the visualization table (or any better replacement of that...). 

Let's get this thing here merged and revisit any revision of visualization-related functionality afterwards?

     (updated) species amounts and corresponding compartment sizes.

4. **Events and finalization**

   At this stage, all remaining state- or time-dependent changes encoded in
   the model are applied, including SBML events. The model state used for the
   evaluation of trigger functions or the execution of any event assignments
   (including events with delays) is the model state after (3).

   The initial model state for the new period at time = ``time``
   (:ref:`v2_experiments_table`) is the model state after all these updates
   have been applied.

5. **Evaluation of observables**

   If measurements exist for the current timepoint, the observables are
   evaluated after all changes have been applied. The resulting values are
   then compared against the corresponding measurements in the measurement table.

.. _v2_experiments_table:

Experiments table
-----------------

The optional experiments table defines a sequence of experimental
conditions (i.e., discrete changes; see :ref:`v2_conditions_table`) applied
to the model.

The experiments table is provided as a tab-separated values (TSV) file with
the following structure:

> +---------------------+-------------------+-----------------+
> | experimentId        | time              | conditionId     |
> +=====================+===================+=================+
> | PETAB_ID            | NUMERIC or '-inf' | CONDITION_ID    |
> +---------------------+-------------------+-----------------+
> | timecourse_1        |                 0 | condition_1     |
> +---------------------+-------------------+-----------------+

Reviewer 3: To match the conditions table section

```suggestion
+---------------------+-------------------+-----------------+
| experimentId        | time              | conditionId     |
+=====================+===================+=================+
| PETAB_ID            | NUMERIC or '-inf' | CONDITION_ID    |
+---------------------+-------------------+-----------------+
| e.g.                |                   |                 |
+---------------------+-------------------+-----------------+
| timecourse_1        |                 0 | condition_1     |
+---------------------+-------------------+-----------------+
```

| timecourse_1        |                10 | condition_2     |
+---------------------+-------------------+-----------------+
| timecourse_1        |               250 | condition_3     |
+---------------------+-------------------+-----------------+
| patient_3           |              -inf | condition_1     |
+---------------------+-------------------+-----------------+
| patient_3           |                 0 | condition_2     |
+---------------------+-------------------+-----------------+
| intervention_effect |               -20 | no_lockdown     |
+---------------------+-------------------+-----------------+
| intervention_effect |                20 | mild_lockdown   |
+---------------------+-------------------+-----------------+
| intervention_effect |                40 | severe_lockdown |
+---------------------+-------------------+-----------------+

The order of the rows is arbitrary, but specifying the rows in ascending order
of time may improve human readability.

Detailed field description
~~~~~~~~~~~~~~~~~~~~~~~~~~

> The time courses table has three mandatory columns ``experimentId``,

Reviewer 3: ```suggestion
The experiment table has three mandatory columns ``experimentId``,
```

Reviewer 5: experiment*s*

Reviewer 3: "Experiment table" in the figure, but fine for me

Reviewer 2: I think that is inconsistent throughout the documentation, library, ...
Let's address that separately and get this monster merged and out of the way.

``time``, and ``conditionId``:

- ``experimentId`` [PETAB_ID, REQUIRED]

  A unique identifier for the experiment, referenced by the ``experimentId``
  column in the :ref:`v2_measurements_table`.

- ``time``: [NUMERIC or ``-inf``, REQUIRED]

  The time at which the specified condition will become active, in the time
  unit specified in the model. ``-inf`` indicates that the respective
  condition will be simulated until a steady state is reached.

  If the simulation of a condition with steady state fails to reach a steady state,
  and the condition is required for the evaluation of simulation at
  measurement points, the evaluation of the model is not well-defined.
  In such cases, PEtab interpreters should notify the user, for example, by
  returning ``NaN`` or ``inf`` values for the objective function.
  PEtab does not specify a numerical criterion for steady states.
  Any event triggers defined in the model must also be checked during this
  pre-simulation.

>   Multiple conditions cannot be applied simultaneously within a single
>   experiment; each ``time`` values must be unique for each ``experimentId``.

Reviewer 3: ```suggestion
  Multiple conditions cannot be applied simultaneously within a single
  experiment; each ``time`` value must be unique for each ``experimentId``.
```

>   ``time`` will override any initial time specified in the model,
>   except in the case of ``time`` =  `ìnf``, in which case the model-specified
>   time will be used (or 0, if the model does not explicitly specify an initial
>   time).

Reviewer 3: ```suggestion
  ``time`` will override any initial time specified in the model,
  except in the case of ``time`` =  ``-ìnf``, in which case the model-specified
  time will be used (or 0, if the model does not explicitly specify an initial
  time).
```

> - ``CONDITION_ID``:
>   [PETAB_ID or empty, REQUIRED, REFERENCES(condition.conditionId)]

Reviewer 3: ```suggestion
- ``conditionId``:
  [CONDITION_ID or empty, REQUIRED, REFERENCES(condition.conditionId)]
```

Reviewer 3: Hm not sure if this is correct. I guess it should be "PETAB_ID", but the table above it says "CONDITION_ID".

The PEtab data format specifies a parameter estimation problem using a number
of text-based files (
`Tab-Separated Values (TSV) <https://www.iana.org/assignments/media-types/text/tab-separated-values>`_)
(Figure 2), i.e.:

- A model

- A measurement file to fit the model to [TSV]

- (optional) A conditions file specifying model inputs and condition-specific
  parameters [TSV]

- (optional) An experiments file, which describes a sequence of different
  experimental conditions that are applied to the model [TSV]

- An observables file specifying the observation model [TSV]

- A parameters file specifying estimatable parameters and related information
  [TSV]

> - A grouping file that lists all of the files and provides additional information

Reviewer 2: In terms of importance, yes. But it will be unclear which files we are referring to.

  type of the model.

- ``operationType`` [STRING, REQUIRED]

  How the target value is to be interpreted. Allowed values are:

  - ``setCurrentValue``: The current value of the target is initialized to the
    value of ``targetValue`` when the respective condition is applied.
    The target must be a constant target or a differential target.

    Any symbols in ``targetValue`` are interpreted as their value at the time
    just before this change is applied (similar
    to an SBML event with ``useValuesFromTriggerTime=True``; in case of the
    first period of a timecourse, they will be interpreted as their initial
    values as defined in the model).

    **TODO** update

  - ``noChange``: This value is used to indicate that no change is associated
    with this condition. This is useful in cases where an explicit condition
    ID is required, but no change is to be applied. (E.g., if pre-simulation
    up to a steady state is to be performed using the default values of the
    model, followed by a different condition.)
    ``targetId`` and ``targetValue`` have to be empty in this case.
    For any ``conditionId`` with ``operationType=noChange``, there must not be

>     any other row with the same ``conditionId`` in the conditions table.

Reviewer 2: @FFroehlich suggested to completely remove `operationType` because after the recent simplification.  

`noChange` would then be specified by an empty conditionID in the experiments table.

Works for me.

@PEtab-dev/petab-editors : What are your thoughts?

  (:ref:`experiments_table`)
* Support for math expressions in the condition table (:ref:`conditions_table`)
* Clarification and specification of various previously underspecified aspects
  (:ref:`math_expressions`, overriding values in the condition table, etc.)
* Support for extensions
* Observable IDs are now allowed to be used in observable/noise formulas
  (:ref:`observables_table`)

.. _model:

Model definition
----------------

PEtab 2.0.0 is agnostic of specific model formats. A model file is referenced
in the PEtab problem description (YAML) via its file name or a URL.

.. _conditions_table:

Conditions table
----------------

The conditions table specifies changes to parameters, initial or intermediate values of
species, and compartments, during specific time periods. The start and end of the time
period is defined in the :ref:`experiments_table`, where multiple conditions can be
combined to specify time courses or experiments that span multiple time periods. Any
changes applied to the model will remain in effect until they are changed by a condition
applied at a later time period.

This is specified as a tab-separated value file in the following way:

> **TODO: keep conditionName or not?**

Reviewer: conditionName I think can be kept as optional as it can help with nicer plotting.

Reviewer 2: The problem is that now there may be several rows per condition, which makes it not so ideal to specify any condition names there. Would you still keep it that table? Would all rows for that condition have to have the same value, or should there only be one non-empty entry per condition?

Reviewer 2: To be continued at https://github.com/PEtab-dev/PEtab/pull/581/files#r1993877029; marking this as resolved.

    model that are not targets of rate rules or assignment rules.

- ``operationType`` [STRING, REQUIRED]

  How the target value is to be interpreted. Allowed values are:

  - ``setCurrentValue``: The current value of the target is initialized to the
    value of ``targetValue`` when the respective condition is applied.
    Any symbols in ``targetValue`` are interpreted as their value at the time
    just before this change is applied (similar
    to an SBML event with ``useValuesFromTriggerTime=True``; in case of the
    first period of a timecourse, they will be interpreted as their initial
    values as defined in the model).
    The target must be a constant target or a differential target.

  - ``setRate``: The time-derivative of the target is set to ``targetValue``
    (symbolically; the expression is not evaluated at the time the condition is
    applied).
    The target must be a differential target.

  - ``addToRate``: The target's rate is increased by ``targetValue``
    (symbolically).
    The target must be a differential target.

  - ``setAssignment``: The target is set to the value of ``targetValue``

>     (symbolically).

Reviewer 5: No, this is discussed after Differential/Algebraic/Constant targets are introduced. To my understanding, structural alterations are not allowed.

>     (symbolically).

Reviewer 2: We might have different understandings of "structural alterations". Changing some algebraic expression `e := x` to `e := y` would be allowed, but changing some differential state to some algebraic expression would not be allowed. In how much that requires building different models is probably very much implementation dependent.

>     (symbolically).

Reviewer: > We might have different understandings of "structural alterations". Changing some algebraic expression e := x to e := y would be allowed, but changing some differential state to some algebraic expression would not be allowed. In how much that requires building different models is probably very much implementation dependent.

Thanks for clarification, just a last question to fully understand. So if we have a parameter given via an assignment rule; `k = k1 * k2` would it then be allowed to change `k` via `setAssignment`?

>     (symbolically).

Reviewer 2: Yes, that was to my understanding the (only) use case for `setAssignment`.

  If the model has a concept of species and a species ID is provided, its
  value is interpreted as amount or concentration in the same way as anywhere
  else in the model.

**TODO**: clarify order of evaluation

**Examples:** **TODO**

    -  given a timecourse ``0:condition1;10:condition2`` and two constant
       model parameters ``par1``, ``par2`` and the two conditions:

      - ``condition1``: {``par1=0.1``, ``par2=0.2``}
      - ``condition2``: {``par1=par2``, ``par2=par1``}

      This is okay, since no circular dependencies exist: ``par1 = 0.2``, ``par2=0.1``

    - given a ``timecourse 0:condition1`` and two model parameters
      ``par1``, ``par2`` with only a single condition:

      - ``condition1``: {``par1=par2``, ``par2=par1``}

      This is not allowed, in the first condition of the timecourse ``par1``, ``par2``
      cannot be used in the right-hand side of the assignment

    - Given a condition: ``condition1``: {``par1=par3``, ``par2=2*par3``}

      This is allowed.

    Condition changes should be implemented to respect the dependency
    graph between model components:

    - When a condition changes quantity ``A`` and ``B``, and ``B`` is dependent on

>       ``A``, the change in quantity A should be applied first such that the

Reviewer 3: > This is not how SBML work right?
> 
> If I remember correctly if both `A` and `B` are assigned something in an event, basically the pre-event `A` and `B` values are used unless priority is specified.

I'm not sure, all I see in the SBML spec is (regarding simultaneously-occurring events)
> A given simulator may use any algorithm to choose an order as long as every event is executed exactly once.

This choice of compartment priority (and more generally to respect the dependency graph) is partly based on discussions with @matthiaskoenig last year -- perhaps @matthiaskoenig can say more here?

>       ``A``, the change in quantity A should be applied first such that the

Reviewer 5: There is a difference between simultaneously occuring events and simultaneous assignments during the same event. I would treat assignments via conditions the same as a single event, i.e. not subject to priority or useValuesFromTriggerTime attributes. I would propose to not follow the dependency graph (and rather simply use pre-assignment values) with the exception of compartments. As done with SBML Event Assignments (even though this isn't crystal clear in the sbml spec).

For example it might make sense to average two species A = B = (A + B)/2. Due to circularity, this assignment is not possible when following the dependency graph, but straightforward with pre-change values. Following of dependencies can always be reproduces through substitution.

>       ``A``, the change in quantity A should be applied first such that the

Reviewer 4: We want 2 different behaviors. These are managed in SBML events via the ` useValuesFromTriggerTime` flag. So we need probably this flag on the timecourse to ensure both scenarios are possible (see also comment below)

>       ``A``, the change in quantity A should be applied first such that the

Reviewer 2: I'd vote for @FFroehlich's suggestion here.

  - ``STRING``: Any string
  - ``NUMERIC``: Any number excluding ``NaN`` / ``inf`` / ``-inf``
  - ``MATH_EXPRESSION``: A mathematical expression according to the
    `PEtab math expression syntax <math_expressions>`_.
  - ``PETAB_ID``: A string that is a valid PEtab ID
  - ``NON_ESTIMATED_ENTITY_ID``: A string that is a valid PEtab ID and refers
    to an entity that has some associated numeric value and is not estimated,
    e.g., a model parameter, parameter table parameter, or a species in the
    model.

Changes from PEtab 1.0.0
------------------------

PEtab 2.0.0 is a major update of the PEtab format. The main changes are:

* Support for non-SBML models (:ref:`model`)
* Changed condition table format (wide -> long) (:ref:`conditions_table`)
* ``simulationConditionId`` and ``preequilibrationConditionId`` in the
  :ref:`measurements_table` are replaced by ``experimentId`` and a more
  flexible way for defining experiments / time courses
  (:ref:`experiments_table`)
* Support for math expressions in the condition table (:ref:`conditions_table`)
* Clarification and specification of various previously underspecified aspects

>   (:ref:`math_expressions`, overriding values in the condition table, etc.)

Reviewer 2: To be updated once all changes here have been finalized.

* Support for extensions
* Observable IDs are now allowed to be used in observable/noise formulas
  (:ref:`observables_table`)

.. _model:

Model definition
----------------

PEtab 2.0.0 is agnostic of specific model formats. A model file is referenced
in the PEtab problem description (YAML) via its file name or a URL.

.. _conditions_table:

Conditions table
----------------

The conditions table specifies changes to parameters, initial values of species
and compartments for specific simulation conditions. Different conditions can
be combined to form time courses or experiment through the
:ref:`experiments_table`. Any changes applied to the model will remain in
effect until they are changed by a subsequent condition.

This is specified as a tab-separated value file in the following way:

**TODO: keep conditionName or not?**

+--------------+------------------+------------------+-----------------+--------------------+
| conditionId  | [conditionName]  | targetId         | operationType   | targetValue        |
+==============+==================+==================+=================+====================+

> | STRING       | [STRING]         | MODEL_ENTITY_ID  | STRING          | MATH_EXPRESSION    |

Reviewer 5: ```suggestion
| PETAB_ID       | [STRING]         | NON_ESTIMATED_ENTITY_ID  | STRING          | MATH_EXPRESSION    |
```

| e.g.         |                  |                  |                 |                    |
+--------------+------------------+------------------+-----------------+--------------------+
| conditionId1 | conditionName1   | modelEntityId1   | setCurrentValue | 0.42               |
+--------------+------------------+------------------+-----------------+--------------------+
| conditionId1 | conditionName1   | modelEntityId2   |                 | 0.42               |
+--------------+------------------+------------------+-----------------+--------------------+
| conditionId2 | ...              | modelEntityId1   |                 | modelEntityId1 + 3 |
+--------------+------------------+------------------+-----------------+--------------------+
| conditionId2 | ...              | someSpecies      | setCurrentValue | 8                  |
+--------------+------------------+------------------+-----------------+--------------------+
| ...          | ...              | ...              |                 | ...                |
+--------------+------------------+------------------+-----------------+--------------------+

> Each line specifies a change of one specific property of a specific entity.

Reviewer 2: This hides the fact that a single entity has different types of associated values (e.g., current value vs. rate) and excludes "symbolic assignment" like `setAssignment`, doesn't it?

Row- and column-ordering are arbitrary, although specifying ``conditionId``
first may improve human readability.

Detailed field description
~~~~~~~~~~~~~~~~~~~~~~~~~~

- ``conditionId`` [PETAB_ID, REQUIRED]

  Unique identifier for the simulation/experimental condition, to be referenced
  by the :ref:`experiments_table`.

- ``conditionName`` [STRING, OPTIONAL] **TODO: remove**

  Condition names are arbitrary strings to describe the given condition.
  They may be used for reporting or visualization.

- ``targetId`` [NON_ESTIMATED_ENTITY_ID, required except for ``operationType=noChange``]

  The ID of the an entity that is to be changed when applying this condition.
  This may an entity defined in the model, or an alias from the mapping
  table. The entity must be uniquely identifiable. Parameters from the
  :ref:`parameters_table` are not allowed here.
  Different restrictions apply depending on the ``operationType`` and the

>   type of the model.

Reviewer 2: > > This may an entity defined in the model, or an alias from the mapping
> > table.
> 
> `NON_ESTIMATED_ENTITY_ID` is already defined, so could move this statement about the mapping table to that definition.

:+1: 

> > The entity must be uniquely identifiable.
> 
> Do you have an example for why it's useful to state this here? Local parameters of SBML reactions might be? Still this is part of the definition of `NON_ESTIMATED_ENTITY_ID` I think, so could be moved there.

Can be moved there. Right, local parameters would be an example.

> > Parameters from the :ref:`parameters_table` are not allowed here.
> 
> Why can't I change the value of a non-estimated parameter that is in the parameter table? Fine for me, but now this also contradicts the definition of `NON_ESTIMATED_ENTITY_ID`, which explicitly supports parameter table parameters.

That wasn't allowed so far, and I would find it very confusing. Unlike SBML-parameters, PEtab parameters were always assumed to be constant throughout time and conditions. I would strongly prefer keeping it that way. Right, currently this is contradictory and `NON_ESTIMATED_ENTITY_ID` needs to be updated.

>   type of the model.

Reviewer 3: > That wasn't allowed so far, and I would find it very confusing. Unlike SBML-parameters, PEtab parameters were always assumed to be constant throughout time and conditions. I would strongly prefer keeping it that way. Right, currently this is contradictory and `NON_ESTIMATED_ENTITY_ID` needs to be updated.

Yes, you're right, makes sense :+1:

    Referred to as `differential targets` below.

  * Entities that are defined in terms of an algebraic assignment, i.e., they
    are not subject to time-derivative information, but are not generally
    constant, e.g., the targets of SBML assignment rules.
    Referred to as `algebraic targets` below.

  * Entities are that are defined in terms of a constant value but may be
    subject to event assignments, e.g., parameters of an SBML model that are
    not targets of rate rules or assignment rules.
    Referred to as `constant targets` below.

- ``operationType`` [STRING, REQUIRED]

  How the target value is to be interpreted. Allowed values are:

  - ``setCurrentValue``: The current value of the target is initialized to the
    value of ``targetValue`` when the respective condition is applied.
    Any symbols in ``targetValue`` are interpreted as their value at the time
    just before this change is applied (similar
    to an SBML event with ``useValuesFromTriggerTime=True``; in case of the
    first period of a timecourse, they will be interpreted as their initial
    values as defined in the model).
    The target must be a constant target or a differential target.

>   - ``setRate``: The time-derivative of the target is set to ``targetValue``

Reviewer 5: Yeah, that sounds like it would be helpful to introduce `targetValue` in favour of having `addTo`, `multiply`, etc.

Why would we need operators like `rateOf`, `rateExpressionOf`? I don't see why one would ever transfer information from rate to value and everything else should be achievable one way or another.

Also, what about using `setValue` in favor of `setCurrentValue`?

>   - ``setRate``: The time-derivative of the target is set to ``targetValue``

Reviewer 2: One use case where one might want to use both the current value and the rate expression could be some piecewise definition of a rate expression where the condition depends on the current state. I'd don't consider super important.

    (symbolically; the expression is not evaluated at the time the condition is
    applied).
    The target must be a differential target.

  - ``addToRate``: The target's rate is increased by ``targetValue``
    (symbolically).
    The target must be a differential target.

  - ``setAssignment``: The target is set to the value of ``targetValue``
    (symbolically).
    The target must be an algebraic target.

  - ``addToAssignment``: The expression assigned to the target is increased by
    ``targetValue`` (symbolically).
    The target must be an algebraic target.

  - ``noChange``: This value is used to indicate that no change is associated
    with this condition. This is useful in cases where an explicit condition
    ID is required, but no change is to be applied. (E.g., if pre-equilibration
    is to be performed using the default values of the model, followed by a
    different condition.)
    ``targetId`` and ``targetValue`` have to be empty in this case.

>     For any conditionId with ``operationType=noChange``, there must not be any
>     other row  with the same ``conditionId`` in the conditions table.

Reviewer 3: ```suggestion
    For any ``conditionId`` with ``operationType=noChange``, there must not be any
    other row with the same ``conditionId`` in the conditions table.
```

  value is interpreted as amount or concentration in the same way as anywhere
  else in the model.

.. _v2_reinitialization_semantics:

Detailed semantics
~~~~~~~~~~~~~~~~~~

All changes defined in the condition table are applied in five consecutive
phases, following the logic of a single SBML event assignment.

1. **Evaluation of** ``targetValues``

   ``targetValues`` are first evaluated using the *current* values of all
   variables in the respective expression.

   * For the initial time period of an experiment, *current* values are
     determined by the initial conditions defined in the model.

   * For subsequent time periods, the *current* values are defined by the
     simulation results at the end of the preceding time period.

2. **Assignment of the evaluated ``targetValues`` to their targets**

   All evaluated ``targetValues`` are simultaneously assigned to their
   corresponding targets.
   It is invalid to apply more than one assignment to a single target
   simultaneously.

   * This means that values of algebraic entities assigned by, e.g., SBML assignment rules
     or PYSB expressions will need to be re-evaluated.

>    * Notably, for SBML models — where concentrations are always treated as
>      derived (algebraic) variables — species with ``hasOnlySubstanceUnits=false`` and
>      simultaneous changes to concentration and compartment size in the
>      condition table will have their amount and compartment size updated
>      first, with the resulting concentration computed as the fraction of
>      amount to compartment size. For further details, refer to SBML semantic
>      test suite case
>      `01779 <https://github.com/sbmlteam/sbml-test-suite/blob/7ab011efe471877987b5d6dcba8cc19a6ff71254/cases/semantic/01779/01779-model.m>`_.

Reviewer 5: ```suggestion
   * These assignments respect language specific interpretations of what variables are *atomic* and which are *derived* or *algebraic*. Most notably, SBML considers species amounts and compartment size but **not** concentrations to be *atomic*. This means that for any entry that has a ``targetId`` that refers to a species with ``hasOnlySubstanceUnits=false`` will have it's ``targetValue`` (interpreted as concentration) converted to amounts using the **current** compartment size and then applied to the ``targetId``s amount. For further details, refer to SBML semantic
     test suite case `01779 <https://github.com/sbmlteam/sbml-test-suite/blob/7ab011efe471877987b5d6dcba8cc19a6ff71254/cases/semantic/01779/01779-model.m>`_.
```

3. **Update of Derived Variables**

>    After the assignment of the target values, all derived variables are
>    updated. If the model format has a concept of species and compartments,
>    this includes the update of species concentrations, which are derived from
>    their amount and compartment size.

Reviewer 5: ```suggestion
   After the assignment of the target values, all derived variables, i.e. algebraic entities, are
   updated.
   
   * These updates respect language specific interpretations of which variables are *atomic* or *derived*/*algebraic*. For example, SBML considers concentrations to be *derived*, meaning that species concentrations will be recomputed based on (updated) values of species amounts and corresponding compartment sizes.
```

- All model entities, column names and row names are case-sensitive
- Fields in "[]" are optional and may be left empty.
- The following data types are used in descriptions below:

  - ``STRING``: Any string
  - ``NUMERIC``: Any number excluding ``NaN`` / ``inf`` / ``-inf``
  - ``MATH_EXPRESSION``: A mathematical expression according to the
    `PEtab math expression syntax <math_expressions>`_.
  - ``PETAB_ID``: A string that is a valid PEtab ID
  - ``NON_ESTIMATED_ENTITY_ID``: A string that is a valid PEtab ID and refers
    directly, or via the mapping table, to an entity that has some associated
    numeric value (e.g. a model parameter, parameter table parameter,
    or a species in the model) and is not listed in the

>     :ref:`v2_parameters_table` (including ``estimate=0`` parameters).

Reviewer 5: If we move this bit below the introduction of differential/constant entities we could simply say differential and constant entities that are not assigned in the parameters table, independent of the assigned value in the estimate column.

  (:ref:`v2_math_expressions`, overriding values in the condition table, etc.)
* Support for extensions
* Observable IDs are now allowed to be used in observable/noise formulas
  (:ref:`v2_observables_table`)

.. _v2_model:

Model definition
----------------

PEtab 2.0.0 is agnostic of specific model formats. A model file is referenced
in the PEtab problem description (YAML) via its file name or a URL.

In the following, we distinguish between three types of entities:

* **Differential entities**: Entities that are defined in terms of a
  time-derivative, e.g., the targets of SBML rate rules or species that
  change by participating in reactions (educts or products).

* **Algebraic entities**: Entities that are defined in terms of an algebraic
  assignment, i.e., they are not subject to time-derivative information, but
  are not generally constant, e.g., the targets of SBML assignment rules.

* **Constant entities**: Entities are that are defined in terms of a constant
  value but may be subject to event assignments, e.g., parameters of an SBML
  model that are not targets of rate rules or assignment rules.

.. _v2_conditions_table:

> Conditions table

Reviewer 2: I'd prefer introducing the independent tables before they are referenced by other tables (conditions, experiments, observables, measurements, parameters).

----------------

The conditions table specifies changes to parameters, initial or intermediate

> values of species, and compartments, during specific time periods.

Reviewer 2: I find it a bit difficult. I already find the term "condition" a bit problematic, which basically describes a set of discrete changes. A "period" is the time between applying some condition and applying the next condition, or the end of the simulation. They aren't really interchangeable.

> values of species, and compartments, during specific time periods.

Reviewer 2: Tried to do that...

The start and end of the time period is defined in the
:ref:`v2_experiments_table`, where multiple conditions can be combined to
specify time courses or experiments that span multiple time periods.

> Any changes applied to the model will remain in effect until they are changed

Reviewer 5: Probably best to drop it here, there is a more detailed description about model state/structure further below anyways.

by a condition applied at a later time period or by the model itself.

The conditions table is specified as a tab-separated value file in the
following way:

+--------------+-------------------------+--------------------+

> | conditionId  | targetId                | targetValue        |

Reviewer 2: Remove or add examples everywhere?

> | conditionId  | targetId                | targetValue        |

Reviewer: Second this that examples can help digest the standard.

> | conditionId  | targetId                | targetValue        |

Reviewer 2: Agreed, although in the longer run, I'd prefer some self-contained problem-specific HowTos / examples outside this document. I think this will be more useful.

     simulation results at the end of the preceding time period.

2. **Assignment of the evaluated ``targetValues`` to their targets**

   All evaluated ``targetValues`` are simultaneously assigned to their
   corresponding targets.
   It is invalid to apply more than one assignment to a single target
   simultaneously.

   * This means that values of variables assigned by SBML assignment rules
     or PYSB expressions will need to be re-evaluated.

   * Notably, for SBML models — where concentrations are always treated as
     derived variables — species with ``hasOnlySubstanceUnits=false`` and
     simultaneous changes to concentration and compartment size in the
     condition table will have their amount and compartment size updated
     first, with the resulting concentration computed as the fraction of
     amount to compartment size. For further details, refer to SBML semantic
     test suite case
     `01779 <https://github.com/sbmlteam/sbml-test-suite/blob/7ab011efe471877987b5d6dcba8cc19a6ff71254/cases/semantic/01779/01779-model.m>`_.

3. **Update of Derived Variables**

   After the assignment of the target values, all derived variables are

>    updated. If the model format has a concept of species and compartments,

Reviewer 2: It can be combined with the previous step, but I think it's not the same. The one thing is updating the basic model state, and the other one, re-evaluating any dependent constructs. With concentration-species, this is distinction is a bit blurred, though.

I wouldn't say it's wrong for PySB models. If there are no concentrations to update, then this is simple skipped.

>    updated. If the model format has a concept of species and compartments,

Reviewer 5: Ah I see, will make a proposal how to split the description then. I think this will also make it more evident what's actually going on with SBML concentrations/volumes.

   this includes the update of species concentrations, which are derived from
   their amount and compartment size.

4. **Events and Finalization**

   All other state- or time-dependent changes encoded in the model are now
   applied as well. For example, this includes SBML events.
   The model state for the evaluation of any trigger functions for this
   timepoint is the model state after (3).

   The initial model state for the new period at time = ``time``
   (:ref:`v2_experiments_table`) the model state after all of these changes.

> 5. **Evaluation of Observables**

Reviewer 2: I found it more suitable here, but it's probably good to add a reference.

Detailed field description
~~~~~~~~~~~~~~~~~~~~~~~~~~

- ``conditionId`` [PETAB_ID, REQUIRED]

  Unique identifier for the simulation/experimental condition, to be referenced
  by the :ref:`v2_experiments_table`.

- ``conditionName`` [STRING, OPTIONAL] **TODO: remove**

  Condition names are arbitrary strings to describe the given condition.
  They may be used for reporting or visualization.

- ``targetId``
  [NON_ESTIMATED_ENTITY_ID, required except for ``operationType=noChange``]

  The ID of the non-estimated entity that will change.
  Different restrictions apply depending on the ``operationType`` and the
  type of the model.

- ``targetValue`` [MATH_EXPRESSION, REQUIRED]

  The value or expression that will be used to change the target.
  The current value of the target specified under ``targetId`` is initialized
  to the value of ``targetValue`` when the respective condition is applied.

  The target must be a constant target or a differential target.

  If the model has a concept of species and a species ID is provided, its
  value is interpreted as amount or concentration in the same way as anywhere
  else in the model.

TODO Detailed semantics
~~~~~~~~~~~~~~~~~~~~~~~

> At the start of a new period, experimental condition changes are implemented.
> These changes (i) are evaluated with the values of the system state variables
> from immediately before the new period and (ii) are implemented before
> model-specific logic (including other events defined in the model).
> The model-specific logic may result in unexpected system states at the start
> of a new period. For example, if a new period changes both a volume and
> concentration, and the model is defined in SBML, then the concentration will
> change again according to the volume change and will therefore not match the
> specified concentration.
> 
> Any symbols in ``targetValue`` are interpreted as their value at the time
> just before this change is applied (similar
> to an SBML event with ``useValuesFromTriggerTime=True``; in case of the
> first period of a timecourse, they will be interpreted as their initial
> values as defined in the model).
> 
> All changes are applied simultaneously,
> with conservation of mass when volumes change, i.e.
> a volume change implies a concentration change, but not vice versa.

Reviewer 5: ```suggestion
All changes defined in the condition table are applied in two consecutive phases, similar to events in SBML.

1. **Evaluation of ``targetValues``**  
   ``targetValues`` are first evaluated using the *current*values of all variables in the respective expression.  
   - For the initial time period of an experiment, *current* values are determined by the initial conditions defined in the model.  
   - For subsequent time periods, the *current* values are defined by the simulation results at the end of the preceding time period.  

2. **Assignment and Update of Derived Variables**  
   Once evaluated, all ``targetValues`` are simultaneously assigned to their corresponding ``targetId``s, and derived variables are updated accordingly.  
   - This means that values of variables assigned by SBML assignment rules or PYSB expressions will need to be re-evaluated.  
   - Notably, for SBML models — where concentrations are always treated as derived variables — species with ``hasOnlySubstanceUnits=false`` and simultaneous changes to concentration and compartment volume in the condition table will have their amount and compartment volume updated first, with the resulting concentration computed as the fraction of amount to volume. For further details, refer to SBML semantic test suite case  
`01779 <https://github.com/sbmlteam/sbml-test-suite/blob/7ab011efe471877987b5d6dcba8cc19a6ff71254/cases/semantic/01779/01779-model.m>`_.
```

Reviewer 2: Thanks, looks good. I'd only add some explanation similar to the one provided in the test case directly here. I'll commit the suggestion to make it easier for others to propose changes if necessary.

Expected to have the following named columns in any (but preferably this)
order:

+--------------+--------------+-------------+--------------+
| observableId | experimentId | measurement | time         |
+==============+==============+=============+==============+

> | observableId | experimentID | NUMERIC     | NUMERIC\|inf |

Reviewer 5: ```suggestion
| PETAB_ID | PETAB_ID | NUMERIC     | NUMERIC\|inf |
```

Detailed field description
~~~~~~~~~~~~~~~~~~~~~~~~~~

> - ``observableId`` [STRING, REQUIRED, REFERENCES(observables.observableID)]

Reviewer 5: ```suggestion
- ``observableId`` [PETAB_ID, REQUIRED, REFERENCES(observables.observableID)]
```

  Observable ID as defined in the observable table described below.

> - ``experimentId`` [STRING, REQUIRED, REFERENCES(experimentsTable.experimentID)]

Reviewer 5: ```suggestion
- ``experimentId`` [[PETAB_ID], REQUIRED, REFERENCES(experimentsTable.experimentID)]
```

  Experiment ID as defined in the experiments table described below. This column may
  have ``NA`` values, which are interpreted as *use the model as is*.
  This avoids the need for "dummy" conditions and experiments if only a single
  condition is required.

- ``measurement`` [NUMERIC, REQUIRED]

  The measured value in the same units/scale as the model output.

- ``time`` [NUMERIC OR STRING, REQUIRED]

>   Time point of the measurement in the time unit specified in the SBML model,

Reviewer 5: Would consider it to be less confusing if ``inf`` was considered a numeric value here.

Reviewer 2: "Finite numeric value or `inf`"? Since `-inf` would not be allowed?

  The datasetId is used to group certain measurements to datasets. This is
  typically the case for data points which belong to the same observable,

>   the same experiment, the same noise model,

Reviewer 2: That sentence continues on the next line.

  the same observable transformation and the same observable parameters.

>   This grouping makes it possible to use the plotting routines that are

Reviewer 2: The latter. I'd keep that for elsewhere. I don't think we should refer to the petab library here, but that's out of scope.

  provided by the PEtab Python library.

- ``replicateId`` [STRING, OPTIONAL]

  The replicateId can be used to discern replicates with the same
  ``datasetId``, which is helpful for plotting e.g. error bars.

.. _v2_experiments_table:

Experiments table
-----------------

> The optional experiments table describes a sequence of different experimental

Reviewer 4: One aspect I’m unsure about is how simple conditions (not part of a time course) are handled. I assumed that the Experiments table is required and that simple conditions would be treated via a time course with time = 0.

The Measurements table specifies that it requires experimentId ([PETAB_ID], REQUIRED, REFERENCES(experimentsTable.experimentID)), meaning it maps to experiments rather than conditions. So there would be a clarification how the simple cases of only conditions at time = 0 would be handled.

I would strongly prefer that the Experiments table is required. For an optimization problem one would need measurements which are linked to experiments and experiments require conditions. So both tables, conditions and experiments, are required. There can be the special case that both tables are empty in case of just mapping to the model as is without any changes.

Reviewer 2: So far, the experiment table is required, unless a simulation is to be performed with the default model settings.

If there is only a single condition that is described by the model, experimentId for the respective measurements may be empty and the experiment table as well as the conditions table may be omitted.

I'd find it somewhat annoying to create those empty files in that case, but I don't feel very strongly about it.

Other opinions?

Reviewer 4: I would prefer to have the experiments table required and have some way to fill out the conditions and experiments also for the special case of no changes. I.e. Having a row in the experiments table where the conditionId is empty (and an empty condition table). This would allow to use the value of the experimentId (experimentName) in plotting and analysis, ... . It is just a special case where the experiment has no changes.

Reviewer 2: This option already exists, independent of requiring the experiment table. I.e. you can have a row in the experiment table with empty conditionId.

Also some experimentName could be provided there, or elsewhere (to be discussed, see other thread / mapping table).

Reviewer 2: @PEtab-dev/petab-editors / all: Optional or mandatory experiments table?

Reviewer 5: Would prefer making both tables optional over forcing the user to add empty dummies. Should be foolproof as the linter should catch missing experimentIds in the measurements table in case those tables were erroneously omitted.

  i.e. the ``time`` values must be unique for each ``experimentId``.
  The order of the rows is arbitrary, but specifying the rows in ascending order
  of time may improve human readability.

  ``time`` will override any initial time specified in the model.

- ``CONDITION_ID``:
  [PETAB_ID or empty, REQUIRED, REFERENCES(conditions.conditionID)]

  Reference to a condition ID in the condition table that
  is to be applied at the given ``time``, or empty to not apply any changes.

  Note: The time interval in which a condition is applied includes the
  respective starting timepoint, but excludes the starting timepoint of
  the following condition. This means that for an experiment where
  ``condition_A`` is applied at ``time_A`` and later ``condition_B`` is
  applied at ``time_B``, then ``condition_A`` is active
  during the interval ``[time_A, time_B)``. This implies that any event
  assignment that triggers at ``time_B`` will occur *after* ``condition_B`` was
  applied and for any measurements at ``time_B``, the observables will be
  evaluated *after* ``condition_B`` was applied. For further details, refer to

>   :ref:`v2_reinitialization_semantics`.

Reviewer 2: Agreed. Yes, good to clarify that.

  ``datasetId``, which is helpful for plotting e.g. error bars.

.. _experiments_table:

Experiments table
-----------------

The optional experiments table describes a sequence of different experimental
conditions (here: discrete changes) that are applied to the model.

This is specified as a tab-separated value file in the following way:

+---------------------+-------------------+-----------------+

> | experimentId        | time              | conditionId     |

Reviewer 5: While I agree that we can drop the name in the conditions table, I think it would be nice to have an optional name column here.

Reviewer 2: Fine to add `experimentName` here if we drop the condition names. However, if we want to keep condition names, which then would have to be specified in some other place (yaml, annotations table, ...?), I would collect all of that in the same place.

Reviewer 5: In that case I would propose to move all names (including observables etc) to such a table. Would anyways be nice as it would introduce the possibility of, e.g. naming states.

Reviewer 2: Shall we repurpose the mapping table for that? I.e. adding a `name` / `label` column there and allowing empty `modelEntityId`? This could also be a place for adding additional annotations, such as SBO terms.

Alternatively, adding that to the yaml file. I'd avoid another additional table.

Reviewer 4: Personally, I don't have an issue with information being duplicated in long tables. I.e. having the same id and name multiple times. This is probably easier to understand then to moving things to the mapping table.
If we start doing that I would prefer a solution which handles more metadata then just names for ids. E.g. annotations to ontologies, arbitrary information fields, ...
E.g. a nice hashmap which allows to annotate identifiers: 
```
{
    id: "condition1",  # id, table required
    table: "condition",
    name: "My condition"  # additional optional fields
    description: "This condition does ..."
    conditionClass: "class 1 - dosing",
    ...
}
```
For me the mapping table feels like a bad hack and prefer the duplication, or a more generic solution which allows to store additional information which can be used by tools.

Reviewer 2: > If we start doing that I would prefer a solution which handles more metadata then just names for ids. E.g. annotations to ontologies, arbitrary information fields, ...

This would be possible that way. The mapping table can have arbitrary additional columns (`description`, `sboTerm`, whatever you like). I am just not sure if we should worry about those here, as long as there is no defined use of this information. The same is true for `name`, just that I would include that in the specs, since there was a way to specify that in PEtab v1.

> mapping table feels like a bad hack

I'd also prefer not having it, but I don't have a better alternative. It's a compromise we have to make to make this table-based definition work, while keeping the number of tables manageable and avoiding duplication.

Reviewer 4: Sounds good to me. Then I am satisfied with the mapping table approach. 

We should not worry about these here. It would just be important that works in principle. I.e. use the mapping table to store additional information mapped on the ids which tools could use (mainly for visualization and communication).

+=====================+===================+=================+

> | PETAB_ID            | NUMERIC or '-inf' | CONDITION_ID    |

Reviewer 5: NUMERIC or '-inf', see above.

Reviewer 3: Agreed, in the measurements table it is `NUMERIC\|inf` instead

| timecourse_1        |                 0 | condition_1     |
+---------------------+-------------------+-----------------+
| timecourse_1        |                10 | condition_2     |
+---------------------+-------------------+-----------------+
| timecourse_1        |               250 | condition_3     |
+---------------------+-------------------+-----------------+
| patient_3           |              -inf | condition_1     |
+---------------------+-------------------+-----------------+
| patient_3           |                 0 | condition_2     |
+---------------------+-------------------+-----------------+
| intervention_effect |               -20 | no_lockdown     |
+---------------------+-------------------+-----------------+
| intervention_effect |                20 | mild_lockdown   |
+---------------------+-------------------+-----------------+
| intervention_effect |                40 | severe_lockdown |
+---------------------+-------------------+-----------------+

Detailed field description
~~~~~~~~~~~~~~~~~~~~~~~~~~

The time courses table with three mandatory columns ``experimentId``,
``time``, and ``conditionId``:

> - ``experimentId`` [STRING, REQUIRED]

Reviewer 5: ```suggestion
- ``experimentId`` [PETAB_ID, REQUIRED]
```

  Identifier of the experiment. The usual PEtab identifier requirements apply.
  This is referenced by the ``experimentId`` column in the measurement table.

- ``time``: [NUMERIC or  ``-inf``, REQUIRED]

  The time when the condition will become active, in the time unit specified

>   in the model. ``-inf`` indicates pre-equilibration (e.g., for drug

Reviewer 5: ```suggestion
  in the model. ``-inf`` indicates that the respective condition will be simulated until a steady state is reached.
```

Strictly speaking equilibration would only be possible for models that satisfy detailed balance, would therefore prefer to avoid using (incorrect) jargon here.

Reviewer 5: **TODO**, behaviour if there is no steady state?

Reviewer 2: Agreed. Any ideas for a more concise but still correct term?

In case of no steady state: `inf` objective value?

Do we want to specify any steady state criteria? We don't do that for integration accuracy and so on, but it feels problematic to not say anything here. We could suggest something like `dx/dt < 100 * integration_tolerance`?

  treatments, the model would be pre-equilibrated with the no-drug condition).

> - ``CONDITION_ID``: Reference to a condition ID in the condition table that

Reviewer 5: [PETAB_ID, REQUIRED, REFERENCES(conditions.conditionID)]

  is to be applied at the given `time`.

>   Note: The time interval in which a condition is applied includes the
>   respective starting timepoint, but excludes the starting timepoint of
>   the following condition. This means that for an experiment
>   ``[time_A:condition_A; time_B:condition_B]``, ``condition_A`` is active
>   during the interval ``[time_A, time_B)``. This implies that any event
>   assignment that triggers at ``time_B`` will occur *after* ``condition_B`` was
>   applied and for any measurements at ``time_B``, the observables will be
>   evaluated *after* ``condition_B`` was applied.

Reviewer 3: This is the old (not long format) version of the experiments table, so it might be confusing to users

```suggestion
  Note: The time interval in which a condition is applied includes the
  respective starting timepoint, but excludes the starting timepoint of
  the following condition. This means that for an experiment where
  ``condition_A`` is applied at ``time_A`` and later ``condition_B`` is
  applied at ``time_B``, then ``condition_A`` is active
  during the interval ``[time_A, time_B)``. This implies that any event
  assignment that triggers at ``time_B`` will occur *after* ``condition_B`` was
  applied and for any measurements at ``time_B``, the observables will be
  evaluated *after* ``condition_B`` was applied.
```

> Multiple conditions can be applied at the same time point by specifying
> multiple lines with the same ``experimentId`` and ``time`` but different
> ``conditionId``. The order of the conditions is arbitrary **????**.
> 
> **TODO** how to deal with conflicts?

Reviewer 3: Suggestion based on previous conversations, not sure how to word it best though
```suggestion
Multiple conditions can be applied at the same time point by specifying
multiple lines with the same ``experimentId`` and ``time`` but different
``conditionId``. The order of the conditions is arbitrary.

All changes across all simultaneous conditions are applied in order of
dependencies, with conservation of mass when volumes change, i.e.
a volume change implies a concentration change, but not vice versa.

Hence, volume changes are applied before concentration changes in the
same compartment, and changes to parameters are applied before
changes to species that depend on the parameters.
```

  returning ``NaN`` or ``inf`` values for the objective function.
  PEtab does not specify a numerical criterion for steady states.

>   For each ``experimentId`` the ``time`` values must be unique. The order of
>   the rows is arbitrary, but specifying the rows in ascending order of time
>   may improve human readability.

Reviewer 3: ```suggestion
  Multiple conditions cannot be applied simultaneously in a single experiment,
  i.e. the ``time`` values must be unique for each ``experimentId``.
  The order of the rows is arbitrary, but specifying the rows in ascending order
  of time may improve human readability.
```

Detailed field description
~~~~~~~~~~~~~~~~~~~~~~~~~~

> - ``parameterId`` [STRING, REQUIRED]

Reviewer 5: ```suggestion
- ``parameterId`` [PARAMETER_ID, REQUIRED]
```

+=============+=================+=========================+=============+============+==============+==========+=====+

> | STRING      | [STRING]        | log10\|lin\|log         | NUMERIC     | NUMERIC    | NUMERIC      | 0\|1     | ... |

Reviewer 3: ```suggestion
| PETAB_ID      | [STRING]        | log10\|lin\|log         | NUMERIC     | NUMERIC    | NUMERIC      | 0\|1     | ... |
```

(e.g., due to inclusion of blanks, dots, or other special characters),
and thus, would not be allowed in the PEtab problem files.

The TSV file has two mandatory columns, ``petabEntityId`` and
``modelEntityId``. Additional columns are allowed.

+---------------+---------------+---------------+
| petabEntityId | modelEntityId | [name]        |
+===============+===============+===============+
| STRING        | STRING        | STRING        |
+---------------+---------------+---------------+
| reaction1_k1  | reaction1.k1  | reaction1 k1  |
+---------------+---------------+---------------+

Detailed field description
~~~~~~~~~~~~~~~~~~~~~~~~~~

- ``petabEntityId`` [PETAB_ID, REQUIRED]

  A valid PEtab identifier (see :ref:`v2_identifiers`) that is not defined in
  any other part of the PEtab problem.
  This identifier may be referenced in condition, measurement, parameter and
  observable tables, but cannot be referenced in the model itself.

  The ``petabEntityId`` may be the same as the ``modelEntityId``, but it must
  not be used to alias an entity that already has a valid PEtab identifier.
  This restriction is to avoid unnecessary complexity in the PEtab problem
  files.

> - ``modelEntityId`` [STRING, REQUIRED]

Reviewer 5: ```suggestion
- ``modelEntityId`` [STRING or empty, REQUIRED]
```

Detailed field description
~~~~~~~~~~~~~~~~~~~~~~~~~~

> - ``petabEntityId`` [STRING, REQUIRED]

Reviewer 5: ```suggestion
- ``petabEntityId`` [PETAB_ID, REQUIRED]
```

fields specifying files empty and reference the other PEtab problems in the
extension specific fields.

Examples
~~~~~~~~

Examples of the visualization table can be found in the
`Benchmark model collection <https://github.com/Benchmarking-Initiative/Benchmark-Models-PEtab/>`_, for example in the `Chen_MSB2009 <https://github.com/Benchmarking-Initiative/Benchmark-Models-PEtab/tree/master/Benchmark-Models/Chen_MSB2009>`_
model.

.. _v2_mapping_table:

Mapping table
-------------

The mapping table maps PEtab entity IDs to model entity IDs, and may be used
for additional annotations. This file is optional.

This file may be used to provide PEtab-compatible aliases to model entities
whose ID in the model would not be a valid identifier in PEtab
(e.g., due to inclusion of blanks, dots, or other special characters),
and thus, would not be allowed in the PEtab problem files.

The TSV file has two mandatory columns, ``petabEntityId`` and
``modelEntityId``. Additional columns are allowed.

+---------------+---------------+---------------+
| petabEntityId | modelEntityId | [name]        |
+===============+===============+===============+

> | STRING        | STRING        | STRING        |

Reviewer 3: ```suggestion
| PETAB_ID        | STRING        | STRING        |
```

  exactly match the species names in the BNGL-generated network file, and no
  pattern matching will be performed.

> - ``name`` [STRING, OPTIONAL]

Reviewer 5: ```suggestion
- ``name`` [STRING or empty, OPTIONAL]
```

  A human-readable name for the entity. This is optional and may be used for

>   reporting or visualization purposes.

Reviewer 5: Say that if no name is provided it defaults to the petabEntityId?

Reviewer 2: Fine to add, but it's anyways not clear where exactly it would be used.
